File: js/src/vision/vision-conformance.test.ts
// The §9 two-phase vision vectors, asserted as a BINDING contract.
//
// `conformance/cases/vision_orchestration.json` is the cross-port contract for
// vision ORCHESTRATION — emit -> fulfil -> assemble, with only the raw model
// call in the middle per-host (ADR-0005). Vision is an injected model like the
// generator and the embedder, so everything around the call is deterministic and
// ports natively (ADR-0010 tier 1).
// This mirrors `python/tests/conformance/test_vision_orchestration_vectors.py`
// and `golang/vision/vision_conformance_test.go` case for case: the committed
// JSON is read as OPAQUE DATA, never re-derived by calling the code under test,
// and every count is pinned EXACTLY.

import { describe, expect, it } from "vitest";

import { loadCase, loadData } from "../conform/fixtures.js";
import { VISION_DESCRIBE_PROMPT } from "../gen/prompts.js";
import { recordFromMapping, type VisionRecord } from "./describe.js";
import { fulfillVisionRequests, type VisionFulfiller } from "./fulfill.js";
import { buildPendingRequest, type BBox, type PendingVisionRequest } from "./requests.js";
import { buildVisionUnits, type EvidenceUnit, type PartitionPath } from "./units.js";

interface VisionCase {
  document_id: string;
  source_uri: string;
  language: string;
  images: { image_id: string; bytes_b64: string }[];
  emit: PendingVisionRequest[];
  fulfilled: Record<string, VisionRecord>;
  assembled_eus: EvidenceUnit[];
  degrade: {
    fulfilled: Record<string, VisionRecord>;
    assembled_eu_ids: string[];
  };
}

const CASE = loadCase<VisionCase>("vision_orchestration.json");
const PROMPTS = loadData<Record<string, string>>("prompts.json");

/** Pinned EXACTLY. A vector silently dropped from the fixture is a weakened
 *  contract that no per-case assertion can see. */
const EXPECTED_IMAGES = 2;
const EXPECTED_EMITTED = 2;
const EXPECTED_FULFILLED = 2;
const EXPECTED_ASSEMBLED = 2;
const EXPECTED_DEGRADED_UNITS = 1;

const PARTITION: PartitionPath = CASE.assembled_eus[0]!.partition;

function decodeBase64(value: string): Uint8Array {
  // eslint-disable-next-line no-undef
  const binary = atob(value);
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
  return out;
}

/** Drive the fixture's IMAGES (not its expectations) through the real emit
 *  phase, taking the prompt from the port's own generated copy so a paraphrase
 *  fails here instead of quietly agreeing with itself. */
function emitFromFixture(): PendingVisionRequest[] {
  return CASE.images.map((image, index) => {
    const expected = CASE.emit[index]!;
    return buildPendingRequest(
      CASE.document_id,
      image.image_id,
      decodeBase64(image.bytes_b64),
      VISION_DESCRIBE_PROMPT,
      {
        page: expected.source_ref.page,
        bbox: expected.source_ref.bbox as BBox | null,
        sourceUri: CASE.source_uri,
      },
    );
  });
}

/** Replay the fixture's own responses in place of the model, keyed on the
 *  EMITTED image_url — a port that emitted a different data URI would not find
 *  its response here, which is what makes the substitution honest. */
function fixtureFulfiller(failing: ReadonlySet<string> = new Set()): VisionFulfiller {
  const byImageUrl = new Map<string, Record<string, unknown>>();
  for (const emitted of CASE.emit) {
    byImageUrl.set(
      emitted.payload.image_url,
      CASE.fulfilled[emitted.request_id] as unknown as Record<string, unknown>,
    );
  }
  return (imageUrl: string, prompt: string) => {
    if (failing.has(imageUrl)) throw new Error("vision endpoint failed for this image");
    expect(prompt, "the host must be handed the emitted prompt verbatim").toBe(
      VISION_DESCRIBE_PROMPT,
    );
    const mapping = byImageUrl.get(imageUrl);
    if (mapping === undefined) throw new Error(`no fixture response for ${imageUrl}`);
    return mapping;
  };
}

describe("vision orchestration conformance", () => {
  it("pins the vector counts exactly", () => {
    expect(CASE.images).toHaveLength(EXPECTED_IMAGES);
    expect(CASE.emit).toHaveLength(EXPECTED_EMITTED);
    expect(Object.keys(CASE.fulfilled)).toHaveLength(EXPECTED_FULFILLED);
    expect(CASE.assembled_eus).toHaveLength(EXPECTED_ASSEMBLED);
    expect(CASE.degrade.assembled_eu_ids).toHaveLength(EXPECTED_DEGRADED_UNITS);
  });

  it("ships the pinned prompt, in the generated copy and in every payload", () => {
    // The prompt rides in every payload the host POSTs verbatim, so a paraphrase
    // is a different model input — pinned as tier-2 shared data (ADR-0010).
    expect(VISION_DESCRIBE_PROMPT).toBe(PROMPTS["vision_describe"]);
    for (const emitted of CASE.emit) {
      expect(emitted.payload.prompt).toBe(PROMPTS["vision_describe"]);
    }
  });

  it("declares each image's TRUE media type, sniffed from the magic bytes", () => {
    // A port that hardcoded image/png passes the first vector and fails the second.
    const prefixes = emitFromFixture().map((r) => r.payload.image_url.split(";", 1)[0]);
    expect(prefixes).toEqual(["data:image/png", "data:image/jpeg"]);
  });

  it.each(CASE.emit.map((e, i) => [e.request_id, i] as const))(
    "emits %s byte-identically to the fixture",
    (_requestId, index) => {
      expect(emitFromFixture()[index]).toEqual(CASE.emit[index]);
    },
  );

  it("emits no credential — only image content and a prompt", () => {
    // The load-bearing ADR-0005 invariant.
    for (const request of emitFromFixture()) {
      expect(Object.keys(request.payload).sort()).toEqual(["image_url", "prompt"]);
    }
  });

  it.each(Object.keys(CASE.fulfilled))("shapes the record for %s", (requestId) => {
    const expected = CASE.fulfilled[requestId]!;
    expect(
      recordFromMapping(expected.image_id, expected as unknown as Record<string, unknown>),
    ).toEqual(expected);
  });

  it("joins every fulfilled response by request_id", async () => {
    const fulfilled = await fulfillVisionRequests(emitFromFixture(), fixtureFulfiller());
    expect(fulfilled).toEqual(CASE.fulfilled);
  });

  it("assembles the pinned figure evidence units", async () => {
    const requests = emitFromFixture();
    const fulfilled = await fulfillVisionRequests(requests, fixtureFulfiller());
    const units = buildVisionUnits(requests, fulfilled, {
      partition: PARTITION,
      language: CASE.language,
    });
    expect(units).toEqual(CASE.assembled_eus);
  });

  it("degrades when a model call fails, and never fabricates a caption", async () => {
    // The degrade path, driven end-to-end: the first image's model call FAILS.
    // Per-request isolation means the second image still produces its unit, and
    // the first produces none — not an empty caption, not a placeholder.
    const requests = emitFromFixture();
    const failing = new Set([requests[0]!.payload.image_url]);
    const fulfilled = await fulfillVisionRequests(requests, fixtureFulfiller(failing));
    expect(fulfilled).toEqual(CASE.degrade.fulfilled);

    const units = buildVisionUnits(requests, fulfilled, {
      partition: PARTITION,
      language: CASE.language,
    });
    expect(units.map((u) => u.eu_id)).toEqual(CASE.degrade.assembled_eu_ids);
    expect(units.map((u) => u.eu_id)).not.toContain(requests[0]!.request_id);
  });

  it("yields no unit for an empty description", async () => {
    // A model that returns junk (no caption, no description) must also degrade —
    // an EU whose text is empty is a citation to nothing.
    const requests = emitFromFixture();
    const fulfilled = await fulfillVisionRequests(requests, () => ({}));
    expect(Object.keys(fulfilled)).toHaveLength(EXPECTED_FULFILLED);
    expect(
      buildVisionUnits(requests, fulfilled, { partition: PARTITION, language: CASE.language }),
    ).toEqual([]);
  });
});

File: golang/vision/vision_conformance_test.go
package vision

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"reflect"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// The §9 two-phase vision vectors, asserted as a BINDING contract.
//
// conformance/cases/vision_orchestration.json is the cross-port contract for
// vision ORCHESTRATION — emit -> fulfil -> assemble, with only the raw model
// call in the middle per-host (ADR-0005). Vision is an injected model like the
// generator and the embedder, so everything around the call is deterministic and
// ports natively (ADR-0010 tier 1).
// python/tests/conformance/test_vision_orchestration_vectors.py holds the
// reference to exactly this file and js/src/vision/vision-conformance.test.ts
// holds JS; this holds Go, reading the committed JSON as OPAQUE DATA and
// asserting every expectation with counts pinned exactly.

type visionCase struct {
	DocumentID string `json:"document_id"`
	SourceURI  string `json:"source_uri"`
	Language   string `json:"language"`
	Images     []struct {
		ImageID  string `json:"image_id"`
		BytesB64 string `json:"bytes_b64"`
	} `json:"images"`
	Emit         []json.RawMessage          `json:"emit"`
	Fulfilled    map[string]json.RawMessage `json:"fulfilled"`
	AssembledEUs []json.RawMessage          `json:"assembled_eus"`
	Degrade      struct {
		Fulfilled      map[string]json.RawMessage `json:"fulfilled"`
		AssembledEUIDs []string                   `json:"assembled_eu_ids"`
	} `json:"degrade"`
}

// The emitted request, decoded only as far as this test needs to replay it.
type emittedShape struct {
	RequestID string `json:"request_id"`
	Payload   struct {
		Prompt   string `json:"prompt"`
		ImageURL string `json:"image_url"`
	} `json:"payload"`
	SourceRef struct {
		Page *int  `json:"page"`
		BBox *BBox `json:"bbox"`
	} `json:"source_ref"`
}

// Pinned EXACTLY. A vector silently dropped from the fixture is a weakened
// contract that no per-case assertion can see.
const (
	expectedImages        = 2
	expectedEmitted       = 2
	expectedFulfilled     = 2
	expectedAssembled     = 2
	expectedDegradedUnits = 1
)

func loadVisionCase(t *testing.T) visionCase {
	t.Helper()
	var c visionCase
	conform.Case(t, "vision_orchestration.json", &c)
	return c
}

func decodeEmitted(t *testing.T, raw []json.RawMessage) []emittedShape {
	t.Helper()
	out := make([]emittedShape, 0, len(raw))
	for _, item := range raw {
		var shape emittedShape
		if err := json.Unmarshal(item, &shape); err != nil {
			t.Fatalf("decode emitted: %v", err)
		}
		out = append(out, shape)
	}
	return out
}

// jsonEqual compares a Go value's serialization to a committed fixture blob by
// value, not by byte order — key order in a JSON object is not part of the
// contract, but every key and every value is.
func jsonEqual(t *testing.T, got any, want json.RawMessage, label string) {
	t.Helper()
	gotBytes, err := json.Marshal(got)
	if err != nil {
		t.Fatalf("%s: marshal: %v", label, err)
	}
	var gotAny, wantAny any
	if err := json.Unmarshal(gotBytes, &gotAny); err != nil {
		t.Fatalf("%s: re-decode: %v", label, err)
	}
	if err := json.Unmarshal(want, &wantAny); err != nil {
		t.Fatalf("%s: decode fixture: %v", label, err)
	}
	if !reflect.DeepEqual(gotAny, wantAny) {
		t.Fatalf("%s mismatch:\n got: %s\nwant: %s", label, gotBytes, want)
	}
}

// emitFromFixture drives the fixture's IMAGES (not its expectations) through the
// real emit phase, taking the prompt from the port's own embedded copy so a
// paraphrase fails here instead of quietly agreeing with itself.
func emitFromFixture(t *testing.T, c visionCase) []PendingRequest {
	t.Helper()
	expected := decodeEmitted(t, c.Emit)
	requests := make([]PendingRequest, 0, len(c.Images))
	for i, image := range c.Images {
		data, err := base64.StdEncoding.DecodeString(image.BytesB64)
		if err != nil {
			t.Fatalf("decode image %s: %v", image.ImageID, err)
		}
		sourceURI := c.SourceURI
		requests = append(requests, BuildPendingRequest(
			c.DocumentID, image.ImageID, data, DescribePrompt(),
			BuildRequestOptions{
				Page:      expected[i].SourceRef.Page,
				BBox:      expected[i].SourceRef.BBox,
				SourceURI: &sourceURI,
			},
		))
	}
	return requests
}

// fixtureFulfiller replays the fixture's own responses in place of the model,
// keyed on the EMITTED image_url — a port that emitted a different data URI
// would not find its response here, which is what makes the substitution honest.
func fixtureFulfiller(t *testing.T, c visionCase, failing map[string]bool) Fulfiller {
	t.Helper()
	byImageURL := map[string]map[string]any{}
	for _, emitted := range decodeEmitted(t, c.Emit) {
		var mapping map[string]any
		if err := json.Unmarshal(c.Fulfilled[emitted.RequestID], &mapping); err != nil {
			t.Fatalf("decode fulfilled %s: %v", emitted.RequestID, err)
		}
		byImageURL[emitted.Payload.ImageURL] = mapping
	}
	return func(imageURL, prompt string) (map[string]any, error) {
		if failing[imageURL] {
			return nil, errors.New("vision endpoint failed for this image")
		}
		if prompt != DescribePrompt() {
			t.Fatalf("the host must be handed the emitted prompt verbatim")
		}
		mapping, ok := byImageURL[imageURL]
		if !ok {
			t.Fatalf("no fixture response for emitted image_url %q", imageURL)
		}
		return mapping, nil
	}
}

func fixturePartition(t *testing.T, c visionCase) PartitionPath {
	t.Helper()
	var eu struct {
		Partition PartitionPath `json:"partition"`
	}
	if err := json.Unmarshal(c.AssembledEUs[0], &eu); err != nil {
		t.Fatalf("decode partition: %v", err)
	}
	return eu.Partition
}

func TestVisionVectorCountsArePinned(t *testing.T) {
	c := loadVisionCase(t)
	if len(c.Images) != expectedImages {
		t.Fatalf("images: got %d, want %d", len(c.Images), expectedImages)
	}
	if len(c.Emit) != expectedEmitted {
		t.Fatalf("emit: got %d, want %d", len(c.Emit), expectedEmitted)
	}
	if len(c.Fulfilled) != expectedFulfilled {
		t.Fatalf("fulfilled: got %d, want %d", len(c.Fulfilled), expectedFulfilled)
	}
	if len(c.AssembledEUs) != expectedAssembled {
		t.Fatalf("assembled_eus: got %d, want %d", len(c.AssembledEUs), expectedAssembled)
	}
	if len(c.Degrade.AssembledEUIDs) != expectedDegradedUnits {
		t.Fatalf("degrade units: got %d, want %d", len(c.Degrade.AssembledEUIDs), expectedDegradedUnits)
	}
}

func TestVisionPromptIsThePinnedPrompt(t *testing.T) {
	// The embedded copy must equal the canonical conformance/prompts.json…
	var prompts map[string]string
	conform.Data(t, "prompts.json", &prompts)
	if DescribePrompt() != prompts["vision_describe"] {
		t.Fatalf("embedded vision prompt drifted from conformance/prompts.json")
	}
	// …and the prompt actually pinned inside every emitted payload.
	c := loadVisionCase(t)
	for _, emitted := range decodeEmitted(t, c.Emit) {
		if emitted.Payload.Prompt != prompts["vision_describe"] {
			t.Fatalf("%s: emitted prompt is not the pinned prompt", emitted.RequestID)
		}
	}
}

func TestEmittedRequestsMatchTheFixture(t *testing.T) {
	c := loadVisionCase(t)
	requests := emitFromFixture(t, c)
	if len(requests) != len(c.Emit) {
		t.Fatalf("emitted %d requests, fixture pins %d", len(requests), len(c.Emit))
	}
	for i, request := range requests {
		jsonEqual(t, request, c.Emit[i], "emit["+request.RequestID+"]")
	}
}

func TestEmittedPayloadDeclaresTheTrueMediaType(t *testing.T) {
	// The fixture's value is that the payload declares each image's TRUE type,
	// sniffed from the magic bytes — a port that hardcoded image/png passes the
	// first vector and fails the second.
	c := loadVisionCase(t)
	requests := emitFromFixture(t, c)
	want := []string{"data:image/png;base64,", "data:image/jpeg;base64,"}
	for i, request := range requests {
		if len(request.Payload.ImageURL) < len(want[i]) || request.Payload.ImageURL[:len(want[i])] != want[i] {
			t.Fatalf("request %d: image_url does not start with %q", i, want[i])
		}
	}
}

func TestFulfilJoinsEveryResponseByRequestID(t *testing.T) {
	c := loadVisionCase(t)
	requests := emitFromFixture(t, c)
	fulfilled := FulfillRequests(requests, fixtureFulfiller(t, c, nil))
	if len(fulfilled) != len(c.Fulfilled) {
		t.Fatalf("fulfilled %d, fixture pins %d", len(fulfilled), len(c.Fulfilled))
	}
	for requestID, want := range c.Fulfilled {
		record, ok := fulfilled[requestID]
		if !ok {
			t.Fatalf("missing fulfilled record for %s", requestID)
		}
		jsonEqual(t, record, want, "fulfilled["+requestID+"]")
	}
}

func TestAssembledUnitsMatchTheFixture(t *testing.T) {
	c := loadVisionCase(t)
	requests := emitFromFixture(t, c)
	fulfilled := FulfillRequests(requests, fixtureFulfiller(t, c, nil))
	units := BuildUnits(requests, fulfilled, BuildUnitOptions{
		Partition: fixturePartition(t, c),
		Language:  c.Language,
	})
	if len(units) != len(c.AssembledEUs) {
		t.Fatalf("assembled %d units, fixture pins %d", len(units), len(c.AssembledEUs))
	}
	for i, unit := range units {
		jsonEqual(t, unit, c.AssembledEUs[i], "assembled_eus["+unit.EuID+"]")
	}
}

func TestAFailingModelCallDegradesAndNeverFabricates(t *testing.T) {
	// The degrade path, driven end-to-end: the first image's model call FAILS.
	// Per-request isolation means the second image still produces its unit, and
	// the first produces none — not an empty caption, not a placeholder.
	c := loadVisionCase(t)
	requests := emitFromFixture(t, c)
	failing := map[string]bool{requests[0].Payload.ImageURL: true}
	fulfilled := FulfillRequests(requests, fixtureFulfiller(t, c, failing))

	if len(fulfilled) != len(c.Degrade.Fulfilled) {
		t.Fatalf("degrade fulfilled %d, fixture pins %d", len(fulfilled), len(c.Degrade.Fulfilled))
	}
	for requestID, want := range c.Degrade.Fulfilled {
		record, ok := fulfilled[requestID]
		if !ok {
			t.Fatalf("degrade: missing fulfilled record for %s", requestID)
		}
		jsonEqual(t, record, want, "degrade.fulfilled["+requestID+"]")
	}

	units := BuildUnits(requests, fulfilled, BuildUnitOptions{
		Partition: fixturePartition(t, c),
		Language:  c.Language,
	})
	got := make([]string, 0, len(units))
	for _, unit := range units {
		got = append(got, unit.EuID)
	}
	if !reflect.DeepEqual(got, c.Degrade.AssembledEUIDs) {
		t.Fatalf("degrade eu_ids: got %v, want %v", got, c.Degrade.AssembledEUIDs)
	}
	for _, id := range got {
		if id == requests[0].RequestID {
			t.Fatalf("fabricated a unit for the image whose model call failed")
		}
	}
}

func TestAnEmptyDescriptionYieldsNoUnit(t *testing.T) {
	// A model that returns junk (no caption, no description) must also degrade —
	// an EU whose text is empty is a citation to nothing.
	c := loadVisionCase(t)
	requests := emitFromFixture(t, c)
	fulfilled := FulfillRequests(requests, func(string, string) (map[string]any, error) {
		return map[string]any{}, nil
	})
	if len(fulfilled) != expectedFulfilled {
		t.Fatalf("fulfilled %d, want %d", len(fulfilled), expectedFulfilled)
	}
	units := BuildUnits(requests, fulfilled, BuildUnitOptions{
		Partition: fixturePartition(t, c),
		Language:  c.Language,
	})
	if len(units) != 0 {
		t.Fatalf("an empty description produced %d units, want 0", len(units))
	}
}
